core: add resetConnectBackoff() method to ManagedChannel

ericgribkoff opened the pull request:
This addresses https://github.com/grpc/grpc-java/issues/2169. Currently users of OkHttp may have to wait up to 60 seconds (for DNS resolution to be re-attempted) or longer (if in exponential backoff to reconnect after `TRANSIENT_FAILURE`) for RPCs to succeed after the network connection is re-established. This provides a hook for apps to tell gRPC that the network is available and name resolution and reconnection attempts may occur immediately.

dapengzhang0 commented:
+ @zhangkun83

zhangkun83 commented on core/src/main/java/io/grpc/ManagedChannel.java:
```diff
@@ -104,4 +104,22 @@ public ConnectivityState getState(boolean requestConnection) {
+   * @since 1.8.0
+   */
+  @ExperimentalApi
+  public void networkAvailable() {
```
This method should be named in an imperative tone like the other methods. Maybe "reconnectNow"?

ericgribkoff replied:
Done

zhangkun83 commented on core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:
```diff
@@ -627,6 +631,26 @@ public void run() {
+        new Runnable() {
+          @Override
+          public void run() {
+            if (!nameResolverStarted || shutdown.get()) {
```
nameResolverStarted should be reset to false when nameResolver is shutdown.

You are skipping the subchannels for the case `!nameResolverStarted` because we know that the channel is in idle mode and all subchannels are shut down. Because it won't harm to call the subchannels anyway, I would prefer not relying on that implication, which is not obvious at first glance. This would be easier to reason about:

```java
if (shutdown.get()) {
  return;
}
if (nameResolverStarted) {
  nameResolver.refresh();
}
for (...
...
```

ericgribkoff replied:
Done

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @ericgribkoff isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
The Linux Foundation CLA GitHub bot_

ericgribkoff commented:
I signed it

ericgribkoff commented:
@zhangkun83 Addressed comments, and rebased onto master to fix the merge conflicts that had popped up in the last two weeks

zhangkun83 commented on core/src/main/java/io/grpc/ManagedChannel.java:
```diff
@@ -104,4 +104,22 @@ public ConnectivityState getState(boolean requestConnection) {
+   */
+  @ExperimentalApi
+  public void reconnectNow() {
+    throw new UnsupportedOperationException("Not implemented");
```
Since this is only advisory, it's safe to use no-op as the default implementation.

ericgribkoff replied:
Done

zhangkun83 commented on core/src/main/java/io/grpc/internal/InternalSubchannel.java:
```diff
@@ -253,12 +257,32 @@ public void run() {
+   * Immediately attempt to reconnect if the current state is TRANSIENT_FAILURE. Otherwise this
+   * method has no effect.
+   */
+  void networkAvailable() {
```
Rename this too

ericgribkoff replied:
Oops, fixed

ejona86 commented:
API discussion notes:

ejona86 commented:
API discussion notes
Names:
networkAvailable(). Sort of wrong; should be imperitive. respondToNetworkAvailable()
connectNow(). Doesn't force a connection though. connectNowIfNecessary(). verbose.
reconnectNow() is a bit confusing when we are already connected, since we won't reconnect. reconnectNowIfNecessary()
resetBackoff()/resetConnectBackoff()
ensureConnectivity(). Doesn't force a connect (getState() does that).

Verdict: +5 votes for resetConnectBackoff() "So far the best"

ericgribkoff commented:
Renamed to `resetConnectBackoff()`. Also changed`InternalSubchannel#cancelReconnectTask` and added test coverage to ensure that the backoff timer is actually reset, for the case where the reconnection attempt triggered by `resetConnectBackoff()` itself fails (thanks @zpencer for asking about this case).

ericgribkoff commented:
From @ejona86's comment on #2169:

> The API in #3580 may be a bit annoying for applications to use, as users will have to plumb the notification through their code to all channels. We may consider alternative approaches.

One alternative discussed yesterday was a global `resetAllConnectBackoff()` method that would reset the connect backoffs for all channels created by the app. I am not opposed to this type of global method, but I don't think we need to go down that route until we have user feedback suggesting it's necessary. In gRPC right now, users must not only keep a reference to all open channels when handling network switches, but also keep track of the creation parameters for each one so that the channel can be shutdown and recreated when the network comes back. With this PR, only the reference to the channel is necessary, and this is, I think, strictly an improvement and not *that* onerous - particularly since it seems like the per-channel API method added in this PR does not preclude also offering the `resetAllConectBackoff()` helper method in the future.

Additionally (and this may or may not be a compelling use case), I can imagine situations where it's desirable to reset the connection on some channels but not all. Some local services may become available when the device connects to a local network without internet connectivity, but there'd be no need to also reset the connect backoff for connections to external hosts. This per-channel functionality is much more straightforward with the API here than it would be with a global `resetAllConnectBackoff(exceptTheseChannels)` type of API.

zhangkun83 commented on core/src/test/java/io/grpc/internal/InternalSubchannelTest.java:
```diff
@@ -884,6 +884,64 @@ public ProxyParameters proxyFor(SocketAddress targetServerAddress) {
+
+    internalSubchannel.resetConnectBackoff();
+
+    assertNoCallbackInvoke();
```
Probably want to also verify that no tasks are scheduled on fakeClock.

ericgribkoff replied:
This is already done (for all tests) in the `@After noMorePendingTasks()` method

zhangkun83 commented on core/src/main/java/io/grpc/ManagedChannel.java:
```diff
@@ -104,4 +104,21 @@ public ConnectivityState getState(boolean requestConnection) {
   }
+
+  /**
+   * Notifies the channel that a reconnect attempt should be made. The implementation may attempt to
```
Please also update the wording here. May be remove the first sentence and organize the rest as:
`For subchannels that are in TRANSIENT_FAILURE status, short-circuit the backoff timer and make them reconnect immediatly . May also attempt to invoke {@link NameResolver#refresh}`

ericgribkoff replied:
Done
